In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

In [2]:
gdp_df = pd.read_csv('../data/gdp_percapita.csv', nrows=6868)
gdp_df.head(6)

Country or Area  Year        Value  Value Footnotes
0     Afghanistan  2019  2065.036398              NaN
1     Afghanistan  2018  2033.779002              NaN
2     Afghanistan  2017  2058.383832              NaN
3     Afghanistan  2016  2057.062164              NaN
4     Afghanistan  2015  2068.265904              NaN
5     Afghanistan  2014  2102.385234              NaN

In [3]:
gdp_df.shape 

(6868, 4)

6868 rows, 4 columns 

In [4]:
internet_df = pd.read_csv('../data/internet_use.csv', nrows=4495)
internet_df.head(6)

Country or Area  Year     Value  Value Footnotes
0     Afghanistan  2014  6.390000              NaN
1     Afghanistan  2013  5.900000              NaN
2     Afghanistan  2012  5.454545              NaN
3     Afghanistan  2011  5.000000              NaN
4     Afghanistan  2010  4.000000              NaN
5     Afghanistan  2009  3.550000              NaN

Look at the shape of each DataFrame - how many rows? How many columns?

gdp- 6868 rows and 4 coumns,   
internet -4495 rows and 4 columns 

In [5]:
gdp_df.shape

(6868, 4)

In [6]:
internet_df.shape

(4495, 4)

Take a look at the datatypes for the columns in each DataFrame.,


In [7]:
gdp_df.dtypes

Country or Area     object
Year                 int64
Value              float64
Value Footnotes    float64
dtype: object

In [8]:
internet_df.dtypes

Country or Area     object
Year                 int64
Value              float64
Value Footnotes    float64
dtype: object

Take a look at the last 10 rows of each DataFrame in turn.

In [9]:
gdp_df.head(10)

Country or Area  Year        Value  Value Footnotes
0     Afghanistan  2019  2065.036398              NaN
1     Afghanistan  2018  2033.779002              NaN
2     Afghanistan  2017  2058.383832              NaN
3     Afghanistan  2016  2057.062164              NaN
4     Afghanistan  2015  2068.265904              NaN
5     Afghanistan  2014  2102.385234              NaN
6     Afghanistan  2013  2116.465454              NaN
7     Afghanistan  2012  2075.491747              NaN
8     Afghanistan  2011  1904.559799              NaN
9     Afghanistan  2010  1957.029338              NaN

In [10]:
internet_df.head(10)

Country or Area  Year     Value  Value Footnotes
0     Afghanistan  2014  6.390000              NaN
1     Afghanistan  2013  5.900000              NaN
2     Afghanistan  2012  5.454545              NaN
3     Afghanistan  2011  5.000000              NaN
4     Afghanistan  2010  4.000000              NaN
5     Afghanistan  2009  3.550000              NaN
6     Afghanistan  2008  1.840000              NaN
7     Afghanistan  2007  1.900000              NaN
8     Afghanistan  2006  2.107124              NaN
9     Afghanistan  2005  1.224148              NaN

Drop the `value footnotes` column from both DataFrames. Check that this worked as expected.

In [11]:
gdp_df.columns 

Index(['Country or Area', 'Year', 'Value', 'Value Footnotes'], dtype='object')

In [12]:
del gdp_df['Value Footnotes']

In [13]:
gdp_df.head()

Country or Area  Year        Value
0     Afghanistan  2019  2065.036398
1     Afghanistan  2018  2033.779002
2     Afghanistan  2017  2058.383832
3     Afghanistan  2016  2057.062164
4     Afghanistan  2015  2068.265904

In [14]:
internet_df.columns 

Index(['Country or Area', 'Year', 'Value', 'Value Footnotes'], dtype='object')

In [15]:
del internet_df['Value Footnotes']

In [16]:
internet_df.head()

Country or Area  Year     Value
0     Afghanistan  2014  6.390000
1     Afghanistan  2013  5.900000
2     Afghanistan  2012  5.454545
3     Afghanistan  2011  5.000000
4     Afghanistan  2010  4.000000

8.Change the columns for the GDP Per Capita DataFrame to ‘Country’, ‘Year’, and ‘GDP_Per_Capita’.,

gdp
Country	Year	GDP_Per_Capita
0	Afghanistan	2019	2065.036398

internet
Country	Year	Internet_Users_Pct
0	Afghanistan	2014	6.39

In [17]:
new_name_gdp = gdp_df.rename({'Country or Area':'Country','Year':'Year','Value':'GDP_Per_Capita'}, axis='columns')
new_name_gdp.head(1)

Country  Year  GDP_Per_Capita
0  Afghanistan  2019     2065.036398

9. Change the columns for the Internet Users DataFrame to ‘Country’, ‘Year’, and ‘Internet_Users_Pct’.

In [18]:
new_name_internet = internet_df.rename({'Country or Area':'Country','Year':'Year','Value':'Internet_Users_Pct'}, axis='columns')
new_name_internet.head(1)

Country  Year  Internet_Users_Pct
0  Afghanistan  2014                6.39

In [19]:
new_name_gdp.head(5)

Country  Year  GDP_Per_Capita
0  Afghanistan  2019     2065.036398
1  Afghanistan  2018     2033.779002
2  Afghanistan  2017     2058.383832
3  Afghanistan  2016     2057.062164
4  Afghanistan  2015     2068.265904

10. Merge the two DataFrames to one. Merge **all rows** from each of the two DataFrames. Call the new DataFrame `gdp_and_internet_use`.

In [20]:
gdp_and_internet_use = new_name_gdp.merge(new_name_internet, left_on = 'Country', right_on ='Country')

Look at the first five rows of your new DataFrame to confirm it merged correctly.

In [21]:
gdp_and_internet_use.head()

Country  Year_x  GDP_Per_Capita  Year_y  Internet_Users_Pct
0  Afghanistan    2019     2065.036398    2014            6.390000
1  Afghanistan    2019     2065.036398    2013            5.900000
2  Afghanistan    2019     2065.036398    2012            5.454545
3  Afghanistan    2019     2065.036398    2011            5.000000
4  Afghanistan    2019     2065.036398    2010            4.000000

Look at the last five rows to make sure the data is clean and as expected.

In [22]:
gdp_and_internet_use.tail()

Country  Year_x  GDP_Per_Capita  Year_y  Internet_Users_Pct
105949  Zimbabwe    1990     3324.348171    1997            0.033080
105950  Zimbabwe    1990     3324.348171    1996            0.016790
105951  Zimbabwe    1990     3324.348171    1995            0.007684
105952  Zimbabwe    1990     3324.348171    1994            0.001739
105953  Zimbabwe    1990     3324.348171    1990            0.000000

Subset the combined DataFrame to keep only the data for 2004, 2009, and 2014. Check that this happened correctly.

In [23]:
gdp_and_internet_use1 = gdp_and_internet_use.loc[gdp_and_internet_use['Year_x' and 'Year_y'].isin([2004, 2009, 2014])]
gdp_and_internet_use1.head(5)

Country  Year_x  GDP_Per_Capita  Year_y  Internet_Users_Pct
0   Afghanistan    2019     2065.036398    2014            6.390000
5   Afghanistan    2019     2065.036398    2009            3.550000
10  Afghanistan    2019     2065.036398    2004            0.105809
15  Afghanistan    2018     2033.779002    2014            6.390000
20  Afghanistan    2018     2033.779002    2009            3.550000

Create three new DataFrames, one for 2004, one for 2009, and one for 2014. Give them meaningful names that aren't too long.

In [24]:
gdp_and_internet_use_2004 =gdp_and_internet_use1.loc[(gdp_and_internet_use1['Year_x'] == 2004) & (gdp_and_internet_use1['Year_y'] == 2004)]
gdp_and_internet_use_2004.head()

Country  Year_x  GDP_Per_Capita  Year_y  Internet_Users_Pct
235           Afghanistan    2004     1200.278256    2004            0.105809
595               Albania    2004     7604.838002    2004            2.420388
1240              Algeria    2004    10058.435939    2004            4.634475
1870               Angola    2004     5593.241537    2004            0.464815
2485  Antigua and Barbuda    2004    18745.681534    2004           24.266544

In [25]:
gdp_and_internet_use_2009 =gdp_and_internet_use1.loc[(gdp_and_internet_use1['Year_x'] == 2009) & (gdp_and_internet_use1['Year_y'] == 2009)]
gdp_and_internet_use_2009.head()

Country  Year_x  GDP_Per_Capita  Year_y  Internet_Users_Pct
155           Afghanistan    2009     1758.904043    2009                3.55
485               Albania    2009    10346.864199    2009               41.20
1125              Algeria    2009    10782.361398    2009               11.23
1765               Angola    2009     7643.493749    2009                6.00
2375  Antigua and Barbuda    2009    20046.839017    2009               42.00

In [38]:
gdp_and_internet_use_2014 =gdp_and_internet_use1.loc[(gdp_and_internet_use1['Year_x'] == 2014) & (gdp_and_internet_use1['Year_y'] == 2014)]
gdp_and_internet_use_2014.head()

Country  Year_x  GDP_Per_Capita  Year_y  Internet_Users_Pct
75            Afghanistan    2014     2102.385234    2014                6.39
375               Albania    2014    11623.866679    2014               60.10
1010              Algeria    2014    11512.705405    2014               18.09
1660               Angola    2014     8239.828837    2014               21.26
2265  Antigua and Barbuda    2014    18104.241470    2014               64.00

Which country had the highest percentage of internet users in 2014? What was the percentage? (Try typing the first 3 letters of your DataFrame name and hitting the tab key for auto-complete options).

In [36]:
gdp_and_internet_use_2014_perc = gdp_and_internet_use_2014.set_index('Country')
gdp_and_internet_use_2014_perc.head(1)

Year_x  GDP_Per_Capita  Year_y  Internet_Users_Pct
Country                                                        
Afghanistan    2014     2102.385234    2014                6.39

gdp_and_internet_use_2014_perc['percent'] =(gdp_and_internet_use_2014_perc['Internet_Users_Pct'] / gdp_and_internet_use_2014_perc['Internet_Users_Pct'].sum())*100

gdp_and_internet_use_2014_perc.head()

In [37]:
gdp_and_internet_use_2014_perc.loc[gdp_and_internet_use_2014_perc['Internet_Users_Pct'].idxmax()]

Year_x                 2014.0000
GDP_Per_Capita        49652.1495
Year_y                 2014.0000
Internet_Users_Pct       98.1600
Name: Iceland, dtype: float64

In [40]:
gdp_and_internet_use_2014_perc.loc[gdp_and_internet_use_2014_perc['Internet_Users_Pct'].idxmin()]

Year_x                2014.00000
GDP_Per_Capita        3264.56807
Year_y                2014.00000
Internet_Users_Pct       1.14000
Name: Timor-Leste, dtype: float64